src/cart/crt_hg.c

```diff
@@ -1270,6 +1270,439 @@ crt_hg_req_destroy(struct crt_rpc_priv *rpc_priv)
+
+/* set to 0 to disable string copies -- faster */
+#define RPT_INCLUDE_STRINGS 1
+
+/* set to 0 to enable on servers too */
+#define RPT_CLIENT_ONLY 1
+
+enum rpt_op {
+	RPT_OP_ALLOC = 1,
+	RPT_OP_ADDREF = 2,
+	RPT_OP_DECREF = 3,
+	RPT_OP_FREE = 4,
+	RPT_OP_COMPLETED = 5,
+	RPT_OP_SEND =6,
+};
+
+struct rpt_alloc_data {
+	int size;
+	bool forward;
+	int opc;
+};
+
+struct rpt_ref_data {
+#if RPT_INCLUDE_STRINGS
+	char func_name[RPT_STRING_SIZE_MAX];
+#endif
+	int line;
+	int ref_count;
+};
+
+struct rpt_free_data {
+};
+
+struct rpt_completed_data {
+	int rc;
+};
+
+struct rpt_send_data {
+	int rank;
+	int tag;
+	int hg_rc;
+};
+
+union rpt_data {
+	struct rpt_alloc_data		_alloc;
+	struct rpt_ref_data		_ref;
+	struct rpt_free_data		_free;
+	struct rpt_completed_data	_completed;
+	struct rpt_send_data		_send;
+};
+
+struct rpt_log_entry {
+	struct timeval		tv;
+	enum rpt_op		op;
+	union rpt_data 		rpt_msg;
+	struct rpt_log_entry 	*next;
+	struct rpt_log_entry 	*prev;
+};
+
+struct rpt_entry {
+	struct timeval		tv_start;
+	uint64_t		addr;
+	pthread_mutex_t		log_mutex;
+	struct rpt_log_entry	*log_head;
+	struct rpt_log_entry	*log_tail;
+	int			num_log_entries;
+	struct rpt_entry	*next;
+	struct rpt_entry	*prev;
+};
+
+static struct rpt_entry *rpt_head;
+static struct rpt_entry *rpt_tail;
+static int rpt_num_entries;
+static pthread_mutex_t rpt_mutex = PTHREAD_MUTEX_INITIALIZER;
+
+
+static struct rpt_entry* __get_rpt_entry(struct crt_rpc_priv *rpc_priv)
+{
+	struct rpt_entry *entry = NULL;
+
+	D_MUTEX_LOCK(&rpt_mutex);
+	entry = rpt_head;
+
+	while (entry) {
+		if (entry->addr == (uint64_t)rpc_priv)
+			break;
+
+		entry = entry->next;
+	}
+	D_MUTEX_UNLOCK(&rpt_mutex);
+
+	return entry;
+}
+
+static struct rpt_entry* __create_rpt_entry(struct crt_rpc_priv *rpc_priv)
+{
+	struct rpt_entry *entry = NULL;
+
+	D_ALLOC_PTR(entry);
+
+	if (!entry) {
+		D_ERROR("Failed to alloc entry\n");
+		return entry;
+	}
+
+	entry->addr = (uint64_t)rpc_priv;
+	entry->log_head = NULL;
+	D_MUTEX_INIT(&entry->log_mutex, NULL);
+
+	gettimeofday(&entry->tv_start, NULL);
+
+	D_MUTEX_LOCK(&rpt_mutex);
+
+
+	entry->num_log_entries = 0;
+	entry->next = NULL;
+	entry->prev = NULL;
+
+	if (rpt_head == NULL) {
+		rpt_head = entry;
+		rpt_tail = entry;
+	} else {
+		rpt_head->prev = entry;
+		entry->next = rpt_head;
+		rpt_head = entry;
+	}
+
+	rpt_num_entries++;
+
+	if (rpt_num_entries > RPT_MAX_ENTRIES) {
+		struct rpt_entry	*to_del;
+		struct rpt_log_entry	*log_entry, *tmp_log_entry;
+
+		to_del = rpt_tail;
+
+		if (rpt_tail && rpt_tail->prev) {
+			rpt_tail->prev->next = NULL;
+			rpt_tail = rpt_tail->prev;
+			log_entry = to_del->log_head;
+
+			while (log_entry) {
+				tmp_log_entry = log_entry;
+				log_entry = log_entry->next;
+
+				D_FREE(tmp_log_entry);
+			}
+
+			D_FREE(to_del);
+			rpt_num_entries--;
+		}
+	}
+
+	D_MUTEX_UNLOCK(&rpt_mutex);
+
+	return entry;
+}
+
+
+static struct rpt_log_entry* __create_log_entry(struct rpt_entry* entry)
```

Comment:
(style)  "foo* bar" should be "foo *bar"
(style)  "foo* bar" should be "foo *bar"